### PR TITLE
Some fixes for the DLC module

### PR DESCRIPTION
**Pull request by vigji**
This PR addresses a serious time bottleneck produced by the fact that currently the DLC interface always read through all the linked videos to create the timestamps arrays (line 166: `timestamps = _get_movie_timestamps(video[0], infer_timestamps=infer_timestamps)`), even when external timestamps are supplied. 

This PR simply adds an argument to the `_get_pes_args` function with the option to skip this search, and speeds up conversion from minutes to seconds in case there's both DLC videos and annotations. 

It actually also fixes an old bug of dlc2nwb that was supposed to be fixed, when `_get_movie_timestamps` ask for the Len of a `cv2` reader.

Along the way, it also purges `os` in favour of `pathlib`. Overall, those utility functions could really help some further cleanup, would be happy to do more but testing is currently a bit lacking for all the edge cases and I probably don't won't to risk too much refactoring...

Let me know how this looks!

### CODE REVIEW

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py [hunk 1]
```diff
@@ -269,9 +275,14 @@ def add_subject_to_nwbfile(
     nwbfile : pynwb.NWBFile
         nwbfile with pes written in the behavior module
     """
-    scorer, df, video, paf_graph, dlc_timestamps, _ = _get_pes_args(config_file, h5file, individual_name)
+    timestamps_available = timestamps is not None
+    print("bottle...")
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:279]:
```suggestion

```

removing print (assuming it was for testing purposes)

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py [hunk 2]
```diff
@@ -269,9 +275,14 @@ def add_subject_to_nwbfile(
...
+    scorer, df, video, paf_graph, dlc_timestamps, _ = _get_pes_args(
+        config_file, h5file, individual_name, timestamps_available
+    )
     if timestamps is None:
         timestamps = dlc_timestamps
+    print("...neck!")
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:285]:
```suggestion
```

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py [hunk 3]
```diff
@@ -119,30 +114,35 @@ def _ensure_individuals_in_header(df, dummy_name):
     return df
 
 
-def _get_pes_args(config_file, h5file, individual_name, infer_timestamps=True):
-    if "DLC" not in h5file or not h5file.endswith(".h5"):
+def _get_pes_args(config_file, h5file, individual_name, timestamps_available=False, infer_timestamps=True):
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:117]:
```suggestion
def _get_pes_args(*, config_file: Path, h5file: Path, individual_name: str, timestamps_available: bool =False, infer_timestamps: bool =True):
```

While we're here, let's add a few things

- keyowrd-only arguments
- annotation typing

What do you think?

A docstring would also be nice (numpy style)

**Comment by CodyCBakerPhD** [comment]
As always, thanks for the enhancement

I think we've been aware of this one for a while unfortunately

Some comments above

Can you add a CHANGELOG entry (improvement category)?

Also, should we propagate this to the interface?

**Comment by vigji** [comment]
Thank you for the feedback! Sorry, I forgot to mark it as draft, yes, there was still some dirt around :) 

I will add changes to the changelog and cross-checking everything. 

My current usage does not require propagation to the interface as it will figure by itself the video availability from the DLC config file. The whole setup looks a bit brittle (my current downstream pipeline has utils to copy the DLC config.yaml file in the data folder, and changing it adding the processed video to the list of videos in the config).

I am available for a further cleanup in the future but will need to work with someone else not to depend too much on my specific configuration!

**Comment by vigji** [comment]
I was driven close to complete madness by the following error, now fixed but not understood:

If I do not`from ndx_pose import PoseEstimationSeries, PoseEstimation` in `deelplabcutinterface.py`, conversion would run smoothly with no errors but then no actual `PoseEstimation` containers would be put in the NWB file. Once it is loaded from disk, one just find an empty `NWBINterface` instead.

I can't really follow what happened there, and sorry if I missed related pointers, but it took an incredibly long amount of time for me to figure this out, basically reverting piece by piece the code to old working version with DLC2NWB, so maybe whatever caused it should be stated very explicitly somewhere!

**Comment by vigji** [comment]
Not sure what is the deal with the tests, as ndx-pose is in the requirements

File: requirements-testing.txt [hunk 4]
```diff
@@ -6,3 +6,4 @@ ndx-miniscope
 spikeinterface[qualitymetrics]>=0.100.0
 zarr<2.18.0  # Error with Blosc (read-only during decode) in numcodecs on May 7; check later if resolved
 pytest-xdist
+ndx-pose
```
Comment (CodyCBakerPhD) [inline requirements-testing.txt:9]: Alas, we can not have this here - a fundamental part of the early testing suite (in CI) is to mimic minimal installation patterns, which will necessarily cause an import error for the optional/extra installation of DLC dependencies

Comment (CodyCBakerPhD) [inline requirements-testing.txt:9]:
```suggestion
```

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py [hunk 5]
```diff
@@ -178,13 +193,11 @@ def _write_pes_to_nwbfile(
     exclude_nans,
     pose_estimation_container_kwargs: Optional[dict] = None,
 ):
-    from ndx_pose import PoseEstimation, PoseEstimationSeries
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:181]: Moving this out of local scope, because this is an unexposed private submodule, might be OK...

File: src/neuroconv/datainterfaces/behavior/deeplabcut/deeplabcutdatainterface.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 from pynwb.file import NWBFile
 
+from ._dlc_utils import add_subject_to_nwbfile
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/deeplabcutdatainterface.py:7]: Unfortunately, this line here indirectly makes it exposed since a parent `__init__.py` somewhere along the chain will do a directed import from this public submodule

Meaning anyone working in a minimal installation, or an installation using extras that do not include DLC, will encounter import errors

**Comment by CodyCBakerPhD** [comment]
> I can reproduce the bug if I run the example in the gallery, meaning that it does not depend on my script; but I find no way of having tests failing because of it, even if I remove the PoseEstimation import from the test initial imports, so this funny behavior will remain untested...

Can you provide an exact set of steps (and perhaps commit hashes) for reproducing the bug?

File: src/neuroconv/datainterfaces/behavior/deeplabcut/deeplabcutdatainterface.py (hunk 6 above)
Comment (vigji) [reply src/neuroconv/datainterfaces/behavior/deeplabcut/deeplabcutdatainterface.py]: of, apparently this is not the crucial change as long as imports in `._dlc_utils` happen at the top of the file

**Comment by CodyCBakerPhD** [comment]
Out of curiosity, how are you running the code? via ipython console? In an IDE? Jupyter notebook?

**Comment by vigji** [comment]
Terminal in VSCode! Oh god, is it IDE-dependent? 😂

**Comment by CodyCBakerPhD** [comment]
By terminal, do you mean direct python invocation of a script? (`python some_script.py`)

Or the integrated console window running in some kernel?

**Comment by CodyCBakerPhD** [comment]
I'm having trouble reproducing your issue

If I run

```python
from zoneinfo import ZoneInfo
from pathlib import Path
from datetime import datetime

import numpy as np
from pynwb import NWBHDF5IO
from neuroconv.datainterfaces import DeepLabCutInterface

base_path = Path("/Users/codybaker")
dlc_path = base_path / "GIN" / "behavior_testing_data" / "DLC"

file_path = dlc_path / "m3v1mp4DLC_resnet50_openfieldAug20shuffle1_30000.h5"
config_file_path = dlc_path / "config.yaml"
path_to_save_nwbfile = base_path / "Downloads" / "test_dlc.nwb"

timestamps = np.array([1,2,3])  # stupid, speeds up processing if tstamps provided
interface = DeepLabCutInterface(file_path=file_path, config_file_path=config_file_path,
                                subject_name="ind1", verbose=False)
interface.set_aligned_timestamps(timestamps)
metadata = interface.get_metadata()
session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=ZoneInfo("US/Pacific"))
metadata["NWBFile"].update(session_start_time=session_start_time)
interface.run_conversion(nwbfile_path=path_to_save_nwbfile, metadata=metadata, overwrite=True)

# confirm the file contains the new TimeSeries in acquisition
with NWBHDF5IO(path_to_save_nwbfile, "r") as io:
    read_nwbfile = io.read()
    print(read_nwbfile.processing)
    print(read_nwbfile.processing["behavior"]["PoseEstimation"])
```

I see

```bash
> {'behavior': behavior pynwb.base.ProcessingModule at 0x5869050416
Fields:
  data_interfaces: {
    PoseEstimation <class 'ndx_pose.pose.PoseEstimation'>
  }
  description: processed behavioral data
}

> PoseEstimation ndx_pose.pose.PoseEstimation at 0x5868721776
Fields:
  description: 2D keypoint coordinates estimated using DeepLabCut.
  dimensions: <HDF5 dataset "dimensions": shape (1, 2), type "<u8">
  nodes: <StrDataset for HDF5 dataset "nodes": shape (4,), type "|O">
  original_videos: <StrDataset for HDF5 dataset "original_videos": shape (1,), type "|O">
  pose_estimation_series: {
    ind1_leftear <class 'ndx_pose.pose.PoseEstimationSeries'>,
    ind1_rightear <class 'ndx_pose.pose.PoseEstimationSeries'>,
    ind1_snout <class 'ndx_pose.pose.PoseEstimationSeries'>,
    ind1_tailbase <class 'ndx_pose.pose.PoseEstimationSeries'>
  }
  scorer: DLC_resnet50_openfieldAug20shuffle1_30000
  source_software: DeepLabCut
```

which shows the data was written correctly to the NWB file on first try (using `main`, not this branch)

Can you try a fresh environment and provide even more explicit instructions to reproduce the error, and what you expected to occur differently?

Preferably in a new issue so we can get this PR in without the import changes

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py (hunk 5 above)
Comment (CodyCBakerPhD) [reply src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:181]: See last comment

We would prefer to keep this import scoped if possible, as a matter of practice

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py [hunk 7]
```diff
@@ -58,14 +58,11 @@ def _get_movie_timestamps(movie_file, VARIABILITYBOUND=1000, infer_timestamps=Tr
     n_frames = int(reader.get(cv2.CAP_PROP_FRAME_COUNT))
     fps = reader.get(cv2.CAP_PROP_FPS)
 
+    warnings.warn("Inferring timestamps from video. This might take a while (to speed up, set timestamps)")
```
Comment (h-mayorquin) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:61]: From memory this takes minutes for hours long videos. Is my memory wrong here? If it is not, I rather would not have this warning by default. I am fine with adding something to skip this relative expensive process but I don't like to use warnings as tool to inform about optimizations. We can do that in the docstring.

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py [hunk 8]
```diff
@@ -58,14 +58,11 @@ def _get_movie_timestamps(movie_file, VARIABILITYBOUND=1000, infer_timestamps=Tr
...
     for _ in range(n_frames):
         _ = reader.read()
         timestamps.append(reader.get(cv2.CAP_PROP_POS_MSEC))
 
-    for _ in range(len(reader)):
```
Comment (h-mayorquin) [inline src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:65]: Can you explain why you removed this? You mention it solves a bug, which bug and how.

Comment (vigji) [reply src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:65]: reader object does not have a `__len__` (maybe it had in previous `cv2` versions?). So I replaced it with `n_frames`. I did not understand how both loops ended up there, I suspect a merging issue.

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py (hunk 7 above)
Comment (vigji) [reply src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:61]: I would suggest to keep the warning. Several minutes, up to 15-20 for me (also, depending on video frame rate, and support from which video is read) can feel very wrong if a user is just linking an external video. At least it did to me. Could go in doctstring, but I would personally find it harder to get this piece of info. 

But I do not encounter this issue myself anymore, totally up to you guys to decide on this!

**Comment by vigji** [comment]
> I'm having trouble reproducing your issue
> 
> If I run
> 
> ```python
> from zoneinfo import ZoneInfo
> from pathlib import Path
> from datetime import datetime
> 
> import numpy as np
> from pynwb import NWBHDF5IO
> from neuroconv.datainterfaces import DeepLabCutInterface
> 
> base_path = Path("/Users/codybaker")
> dlc_path = base_path / "GIN" / "behavior_testing_data" / "DLC"
> 
> file_path = dlc_path / "m3v1mp4DLC_resnet50_openfieldAug20shuffle1_30000.h5"
> config_file_path = dlc_path / "config.yaml"
> path_to_save_nwbfile = base_path / "Downloads" / "test_dlc.nwb"
> 
> timestamps = np.array([1,2,3])  # stupid, speeds up processing if tstamps provided
> interface = DeepLabCutInterface(file_path=file_path, config_file_path=config_file_path,
>                                 subject_name="ind1", verbose=False)
> interface.set_aligned_timestamps(timestamps)
> metadata = interface.get_metadata()
> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=ZoneInfo("US/Pacific"))
> metadata["NWBFile"].update(session_start_time=session_start_time)
> interface.run_conversion(nwbfile_path=path_to_save_nwbfile, metadata=metadata, overwrite=True)
> 
> # confirm the file contains the new TimeSeries in acquisition
> with NWBHDF5IO(path_to_save_nwbfile, "r") as io:
>     read_nwbfile = io.read()
>     print(read_nwbfile.processing)
>     print(read_nwbfile.processing["behavior"]["PoseEstimation"])
> ```
> 
> I see
> 
> ```shell
> > {'behavior': behavior pynwb.base.ProcessingModule at 0x5869050416
> Fields:
>   data_interfaces: {
>     PoseEstimation <class 'ndx_pose.pose.PoseEstimation'>
>   }
>   description: processed behavioral data
> }
> 
> > PoseEstimation ndx_pose.pose.PoseEstimation at 0x5868721776
> Fields:
>   description: 2D keypoint coordinates estimated using DeepLabCut.
>   dimensions: <HDF5 dataset "dimensions": shape (1, 2), type "<u8">
>   nodes: <StrDataset for HDF5 dataset "nodes": shape (4,), type "|O">
>   original_videos: <StrDataset for HDF5 dataset "original_videos": shape (1,), type "|O">
>   pose_estimation_series: {
>     ind1_leftear <class 'ndx_pose.pose.PoseEstimationSeries'>,
>     ind1_rightear <class 'ndx_pose.pose.PoseEstimationSeries'>,
>     ind1_snout <class 'ndx_pose.pose.PoseEstimationSeries'>,
>     ind1_tailbase <class 'ndx_pose.pose.PoseEstimationSeries'>
>   }
>   scorer: DLC_resnet50_openfieldAug20shuffle1_30000
>   source_software: DeepLabCut
> ```
> 
> which shows the data was written correctly to the NWB file on first try (using `main`, not this branch)
> 
> Can you try a fresh environment and provide even more explicit instructions to reproduce the error, and what you expected to occur differently?
> 
> Preferably in a new issue so we can get this PR in without the import changes

Ok, will revert this change and retest main when it will be merged (an alternate solution is to just import `PoseEstimation` in my script atm, so not a pressing matter for myself)

File: src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py (hunk 7 above)
Comment (vigji) [reply src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:61]: Ok! I see the point, thank you for explaining further. Actually yes, having the message on the loading bar would be much better! I'll take the morning out here

Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/behavior/deeplabcut/_dlc_utils.py:61]: Thanks!

**Comment by h-mayorquin** [comment]
This is fine for me to merge when the optimization warning is removed and the top-level import is removed.

**Comment by vigji** [comment]
Ok, this should be done

**Comment by CodyCBakerPhD** [comment]
@vigji Looks like some merge conflicts also need to be updated here

**Comment by CodyCBakerPhD** [comment]
After resolving those this should be good to go though

**Review by CodyCBakerPhD — APPROVED** [review]
